kvserverbase: add DisableCountBasedRebalancingIfMMAEnabled

wenyihu6 opened the pull request:
Resolves: https://github.com/cockroachdb/cockroach/issues/152579
Release note: none

---
**kvserverbase: add DisableCountBasedRebalancingIfMMAEnabled**

This commit introduces a new cluster setting,
`DisableCountBasedRebalancingIfMMAEnabled`, which allows us to disable range-count
and lease-count based convergence rebalancing when mma is enabled. These forms
of rebalancing can conflict with mma goals and cause thrashing.

---
**allocator: add option to disable lease-count based rebalancing**

This commit disables lease-count based rebalancing in the lease queue.

Currently, the lease queue uses `ShouldTransferLease` and `PlanOneChange` to select
a target store for transferring a range lease. Candidate stores are chosen based
on store health, lease preferences, lagging replicas, and I/O load, and the
final target is decided using one of: `FollowTheWorkload`, `LeaseCountConvergence`,
or `LoadConvergence`.

`LeaseCountConvergence` is used in two scenarios:
(1) When `FollowTheWorkload` (`shouldTransferLeaseForAccessLocality`) lacks sufficient
  stats or locality info.
(2) `maybeTransferLeaseAwayTarget`: calls into `allocator.TransferLeaseTarget` with
  `LeaseCountConvergence` directly.

This commit disables case (1) by returning early in
`shouldTransferLeaseForLeaseCountConvergence` (used in
`Allocator.TransferLeaseTarget` and `allocator.ShouldTransferLease`). Case (2) is
left intact, since it is less common and less likely to interfere with MMA
decisions.

---
**kvserver: add `BaseScorerOptions`**

This commit introduces `BaseScorerOptions`, a base struct embedded by other
`ScorerOptions` implementations. It provides the default implementation methods,
while specific implementations can override individual methods to apply their
own scoring heuristics.

---
**allocator: add `adjustRangeCountForScoring`**

This commit adds `adjustRangeCountForScoring` to the `ScorerOptions` interface. The
method takes a provided range count, adjusts it according to the scorer’s
heuristic, and returns the result. It is similar to other scoring metrics such
as the convergence score, and will later be used in candidate.compare as an
attribute for breaking ties. The compare method evaluates the relative
difference between two stores and favors the one with fewer ranges as a better
rebalance target. Currently and by default, `adjustRangeCountForScoring` simply
returns the provided range count. Future scorer options will override it with 0
since the heuristics is to not consider range count at all.

---
**allocator: add `BaseScorerOptionsNoConvergence`**

This commit disables range-count based rebalancing.

Currently, when selecting a rebalance target, the allocator constructs an
equivalence class for each existing replica. This class includes all stores that
are not worse than the existing replica in terms of constraint conformance,
diversity score, and disk fullness. At this stage, range count convergence is
not yet considered. The allocator then evaluates each candidate set using
`ScorerOptions`, which populates candidates' attributes such as balance score,
convergence score, and range count with respect to the equivalence class, and
determines the best candidate within each equivalence class.

To disable range count–based rebalancing, this commit introduces a new scorer
option, `BaseScorerOptionsNoConvergence`. This scorer assigns the same score to
all stores, effectively treating candidates within the same equivalence class
the same. It is only used when both MMA and
`DisableCountBasedRebalancingIfMMAEnabled` are enabled.

Note that this new scorer option is applied only in
`ReplicaPlanner.considerRebalance` and `ReplicaPlanner.ShouldPlanChange`. As a
result, range-count based rebalancing is disabled during rebalancing, but range
count is still considered when allocator adds new replicas (such as
`AllocateVoter`), choosing a rebalance target in the old store rebalancer
(`StoreRebalancer.applyRangeRebalance`), and when removing replicas
(`Allocator.RemoveVoter`).

---
**kvserver: add LBRebalancingMultiMetricAndCount**

This commit removes the cluster setting
DisableCountBasedRebalancingIfMMAEnabled, renames LBRebalancingMultiMetric to
LBRebalancingMultiMetricOnly, and introduces a new mode:
LBRebalancingMultiMetricAndCount. To keep the datadriven test output unchanged,
all previous references to mma in asim now use LBRebalancingMultiMetricAndCount,
which matches the prior behavior. I’ll follow up with a new asim mode in future
commits.

wenyihu6 committed:
kvserverbase: add DisableCountBasedRebalancingIfMMAEnabled

This commit introduces a new cluster setting,
`DisableCountBasedRebalancingIfMMAEnabled`, which allows us to disable range-count
and lease-count based convergence rebalancing when mma is enabled. These forms
of rebalancing can conflict with mma goals and cause thrashing.

wenyihu6 committed:
allocator: add option to disable lease-count based rebalancing

This commit disables lease-count based rebalancing in the lease queue if
CountBasedRebalanceDisabled is true.

Currently, the lease queue uses `ShouldTransferLease` and `PlanOneChange` to select
a target store for transferring a range lease. Candidate stores are chosen based
on store health, lease preferences, lagging replicas, and I/O load, and the
final target is decided using one of: `FollowTheWorkload`, `LeaseCountConvergence`,
or `LoadConvergence`.

`LeaseCountConvergence` is used in two scenarios:
(1) When `FollowTheWorkload` (`shouldTransferLeaseForAccessLocality`) lacks sufficient
  stats or locality info.
(2) `maybeTransferLeaseAwayTarget`: calls into `allocator.TransferLeaseTarget` with
  `LeaseCountConvergence` directly.

This commit disables case (1) by returning early in
`shouldTransferLeaseForLeaseCountConvergence` (used in
`Allocator.TransferLeaseTarget` and `allocator.ShouldTransferLease`). Case (2) is
left intact, since it is less common and less likely to interfere with MMA
decisions.

wenyihu6 committed:
kvserver: add `BaseScorerOptions`

This commit introduces `BaseScorerOptions`, a base struct embedded by other
`ScorerOptions` implementations. It provides the default implementation methods,
while specific implementations can override individual methods to apply their
own scoring heuristics.

wenyihu6 committed:
allocator: add `BaseScorerOptionsNoConvergence`

This commit disables range-count based rebalancing in the replicate queue if
CountBasedRebalanceDisabled is true.

Currently, when selecting a rebalance target, the allocator constructs an
equivalence class for each existing replica. This class includes all stores that
are not worse than the existing replica in terms of constraint conformance,
diversity score, and disk fullness. At this stage, range count convergence is
not yet considered. The allocator then evaluates each candidate set using
`ScorerOptions`, which populates candidates' attributes such as balance score,
convergence score, and range count with respect to the equivalence class, and
determines the best candidate within each equivalence class.

To disable range count–based rebalancing, this commit introduces a new scorer
option, `BaseScorerOptionsNoConvergence`. This scorer assigns the same score to
all stores, effectively treating candidates within the same equivalence class
the same. It is only used when both MMA and
`DisableCountBasedRebalancingIfMMAEnabled` are enabled.

Note that this new scorer option is applied only in
`ReplicaPlanner.considerRebalance` and `ReplicaPlanner.ShouldPlanChange`. As a
result, range-count based rebalancing is disabled during rebalancing, but range
count is still considered when allocator adds new replicas (such as
`AllocateVoter`), choosing a rebalance target in the old store rebalancer
(`StoreRebalancer.applyRangeRebalance`), and when removing replicas
(`Allocator.RemoveVoter`).

wenyihu6 committed:
allocator: add `adjustRangeCountForScoring`

This commit adds `adjustRangeCountForScoring` to the `ScorerOptions` interface. The
method takes a provided range count, adjusts it according to the scorer’s
heuristic, and returns the result. It is similar to other scoring metrics such
as the convergence score, and will later be used in candidate.compare as an
attribute for breaking ties. The compare method evaluates the relative
difference between two stores and favors the one with fewer ranges as a better
rebalance target. Currently and by default, `adjustRangeCountForScoring` simply
returns the provided range count. Future scorer options will override it with 0
since the heuristics is to not consider range count at all.

tbg reviewed (commented):
Looks good so far. One general question about the new setting.

> @tbg reviewed 1 of 1 files at r1, 2 of 2 files at r2, 7 of 7 files at r3, 1 of 1 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153220#-:-OZi_IjC9b_YtdvjgQHp:b109emp)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[`pkg/kv/kvserver/kvserverbase/base.go` line 167 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153220#-OZiYmko2pOdARjVZxE-:-OZiYmko2pOdARjVZxE0:bndb19m) ([raw file](https://github.com/cockroachdb/cockroach/blob/95f22f09a4c967a5c8b24d9cf9ae711f697461fe/pkg/kv/kvserver/kvserverbase/base.go#L167)):*
> ```Go
> // to disable replica and lease count based rebalancing if multi-metric
> // allocator is enabled.
> var DisableCountBasedRebalancingIfMMAEnabled = settings.RegisterBoolSetting(
> ```

I'd like to gently push back against an unprompted proliferation of cluster settings. Can we not just default to disabling count-based rebalancing when multi-metric is active? And if we want the combination of multi-metric plus lease count rebalancing for some reason, couldn't we more transparently add a new enum value to _that_ cluster setting (`multi-metric and count`)?


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/kv/kvserver/kvserverbase/base.go` line 167 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153220#-OZiYmko2pOdARjVZxE-:-OZi_g1S6S8_QKX1yvZ-:b-gt1edl) ([raw file](https://github.com/cockroachdb/cockroach/blob/95f22f09a4c967a5c8b24d9cf9ae711f697461fe/pkg/kv/kvserver/kvserverbase/base.go#L167)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I'd like to gently push back against an unprompted proliferation of cluster settings. Can we not just default to disabling count-based rebalancing when multi-metric is active? And if we want the combination of multi-metric plus lease count rebalancing for some reason, couldn't we more transparently add a new enum value to _that_ cluster setting (`multi-metric and count`)?
</blockquote></details>

Good point - agree that we can default this setting to `true`. I like your idea of adding another enum, `multi-metric-and-count`, so I’ll go ahead with that and remove this cluster setting. It will probably change our asim test output. I’ll cherry-pick your thrashing PR and see if this PR improves things.


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
kvserver: add LBRebalancingMultiMetricAndCount

This commit removes the cluster setting
DisableCountBasedRebalancingIfMMAEnabled, renames LBRebalancingMultiMetric to
LBRebalancingMultiMetricOnly, and introduces a new mode:
LBRebalancingMultiMetricAndCount. To keep the datadriven test output unchanged,
all previous references to mma in asim now use LBRebalancingMultiMetricAndCount,
which matches the prior behavior. I’ll follow up with a new asim mode in future
commits.

tbg reviewed (approved):
> @tbg reviewed 11 of 11 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153220)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[`-- commits` line 56 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/153220#-OZspmqe2FF1tYe88_nm:-OZspmqe2FF1tYe88_nn:b-bwjr9r):*
Not quite, it selectively disables range-count based rebalancing when this is requested.
___
*[`pkg/kv/kvserver/kvserverbase/base.go` line 148 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/153220#-OZsqKLoCDYK-UxhpXYk:-OZsqKLoCDYK-UxhpXYl:b-mh5svo) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c064e51d521b7a59a3650093612928dfc6121aa/pkg/kv/kvserver/kvserverbase/base.go#L148)):*
> ```Go
> // uses the multi-metric store rebalancer.
> var LoadBasedRebalancingModeIsMMA = func(sv *settings.Values) bool {
> 	return LoadBasedRebalancingMode.Get(sv) == LBRebalancingMultiMetricOnly ||
> ```

nit: you don't have to call `Get` twice here. This is never going to matter in practice, but if a user flipped from `AndCount` to `Only`, as written the code may erroneously return false (because the first operand may see `AndCount` and the second one may see `Only`).
___
*[`pkg/kv/kvserver/asim/tests/datadriven_simulation_test.go` line 491 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/153220#-OZsqoQ3BPmbe5UZtV-9:-OZsqoQ3BPmbe5UZtV-A:b-fcmbyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c064e51d521b7a59a3650093612928dfc6121aa/pkg/kv/kvserver/asim/tests/datadriven_simulation_test.go#L491)):*
> ```Go
> 							})
> 					},
> 					"mma-only": func(eg *gen.StaticEvents) {
> ```

You could rename this to `mma-and-count` which wouldn't cause too much churn and leaves things in a consistent place.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153220)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @wenyihu6)

___
*[`pkg/kv/kvserver/asim/tests/datadriven_simulation_test.go` line 491 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/153220#-OZsqoQ3BPmbe5UZtV-9:-OZsr7Qf3_UDBWLfio5k:bj81ivs) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c064e51d521b7a59a3650093612928dfc6121aa/pkg/kv/kvserver/asim/tests/datadriven_simulation_test.go#L491)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

You could rename this to `mma-and-count` which wouldn't cause too much churn and leaves things in a consistent place.
</blockquote></details>

Ah, nevermind - this disables the other queues. You're going to eventually flip this mode to not disable the queues and instead use `Only` here, right?


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153220)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @tbg)

___
*[`-- commits` line 56 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/153220#-OZspmqe2FF1tYe88_nm:-OZyrGp90yP3Ccfdnkhs:b1boeiy):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Not quite, it selectively disables range-count based rebalancing when this is requested.
</blockquote></details>

Reworded to clarify that this is only disabled when CountBasedRebalanceDisabled is true. 
___
*[`pkg/kv/kvserver/kvserverbase/base.go` line 148 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/153220#-OZsqKLoCDYK-UxhpXYk:-OZyrRFM3XQt5BEKzXhD:b-jmaext) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c064e51d521b7a59a3650093612928dfc6121aa/pkg/kv/kvserver/kvserverbase/base.go#L148)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

nit: you don't have to call `Get` twice here. This is never going to matter in practice, but if a user flipped from `AndCount` to `Only`, as written the code may erroneously return false (because the first operand may see `AndCount` and the second one may see `Only`).
</blockquote></details>

Good call on the race here, fixed. 
___
*[`pkg/kv/kvserver/asim/tests/datadriven_simulation_test.go` line 491 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/153220#-OZsqoQ3BPmbe5UZtV-9:-OZyibek3c1O_Q629pp8:b-9o1tw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c064e51d521b7a59a3650093612928dfc6121aa/pkg/kv/kvserver/asim/tests/datadriven_simulation_test.go#L491)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Ah, nevermind - this disables the other queues. You're going to eventually flip this mode to not disable the queues and instead use `Only` here, right?
</blockquote></details>

Putting up https://github.com/cockroachdb/cockroach/pull/153439 to flip the settings and will add more tests there as well. I was hoping to make that PR ready for review before merging this one, but I think it will need a bit more work. I will take that PR over the finish line eventually though.


<!-- Sent from Reviewable.io -->
